python/ray/serve/_private/router.py

```diff
@@ -83,62 +125,194 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
         pass
 
 
-class RoundRobinStreamingReplicaScheduler(ReplicaScheduler):
-    """Round-robins requests across a set of actor replicas using streaming calls.
+class PowerOfTwoChoicesReplicaScheduler(ReplicaScheduler):
+    """Chooses a replica for each request using the "power of two choices" procedure.
+
+    Requests are scheduled in FIFO order.
+
+    When a request comes in, two replicas are chosen randomly as candidates. The queue
+    length of each replicas is requested from it via a control message.
 
-    This policy does *not* respect `max_concurrent_queries`.
+    The replica responds with two items: (queue_length, accepted). Only replicas that
+    accept the request are considered; between those, the one with the lower queue
+    length is chosen.
+
+    In the case when neither replica accepts the request (e.g., their queues are full),
+    the procedure is repeated with backoff. This backoff repeats indefinitely until a
+    replica is chosen, so the caller should use timeouts and cancellation to avoid
+    hangs.
+
+    Each request being scheduled may spawn an independent task that runs the scheduling
+    procedure concurrently. This task will not necessarily satisfy the request that
+    started it (in order to maintain the FIFO order). The total number of tasks is
+    capped at (2 * num_replicas).
     """
 
-    def __init__(self, deployment_name: str):
+    # The sequence of backoff timeouts to use when all replicas' queues are full.
+    # The last item in the list is the max timeout and will be used repeatedly.
+    backoff_sequence_s = [0, 0.05, 0.1, 0.15, 0.2, 0.5, 1.0]
+
+    def __init__(
+        self,
+        event_loop: asyncio.AbstractEventLoop,
+        deployment_name: str,
+        backoff_sequence: List[float] = None,
+    ):
+        self._loop = event_loop
         self._deployment_name = deployment_name
-        self._replica_id_set = set()
-        self._replica_iterator = itertools.cycle([])
+
+        # Current replicas available to be scheduled.
+        # Updated via `update_replicas`.
+        self._replica_id_set: Set[str] = set()
+        self._replicas: Dict[str, ReplicaWrapper] = {}
         self._replicas_updated_event = asyncio.Event()
 
-    async def assign_replica(
-        self, query: Query
-    ) -> "ray._raylet.StreamingObjectRefGenerator":
-        replica = None
-        while replica is None:
-            try:
-                replica = next(self._replica_iterator)
-            except StopIteration:
+        self._scheduling_tasks: Set[asyncio.Task] = set()
+        self._pending_assignment_futures: List[asyncio.Future] = []
```

Comment:
Could you use a `deque` here, so `append` and `pop(0)` are both constant?

Reply:
done!